### PR TITLE
Support Multiple Versions of the PagerDuty Events API

### PR DESCRIPTION
**Pull request by orien**
PagerDuty now provide multiple versions of their Events API: [V1] and [V2].

[V1]: https://v2.developer.pagerduty.com/docs/events-api
[V2]: https://v2.developer.pagerduty.com/docs/events-api-v2

As some PagerDuty organisation/service configurations do not support Events API V2, it's important to continue supporting Events API V1 with this gem.

Looking at this interface provided by this gem, I found that it doesn't easily support two versions of the PagerDuty Events API.

```ruby
pagerduty = Pagerduty.new("<integration-key>")         # <Pagerduty:…>
incident = pagerduty.trigger("<incident-description>") # <PagerdutyIncident:…>
incident.acknowledge
incident.resolve
```

Given the arguments to the `trigger`, `acknowledge`, and `resolve` API calls are quite different between [V1] and [V2], it'll be hard to support both with this class structure.

### Proposal

To alleviate this problem, I propose introducing a new mechanism for instantiating a `pagerduty` instance: The `Pagerduty.build` method.

```ruby
pagerduty = Pagerduty.build(integration_key: "<integration-key>",
                            api_version:     "1")      # <Pagerduty::EventsApiV1:…>
incident = pagerduty.trigger("<incident-description>") # <Pagerduty::EventsApiV1::Incident:…>
incident.acknowledge
incident.resolve
```
This provides a way for the client to specify which version of the Events API is desired and returns an instance with the capacity to interact with the chosen Events API version.

This PR only provides an implementation for [V1], but the path is open to provide a [V2] implementation. It'll look something like this:

```ruby
pagerduty = Pagerduty.build(integration_key: "<integration-key>",
                            api_version:     "2") # <Pagerduty::EventsApiV2:…>
incident = pagerduty.trigger(…)                   # <Pagerduty::EventsApiV2::Incident:…>
incident.acknowledge
incident.resolve
```

I'll add this in a subsequent pull request.

### Backwards Compatibility

To ease the adoption of this new version of the `pagerduty` gem, a layer of backwards compatibility has been provided. In this way, the original interface can still be used:

```ruby
pagerduty = Pagerduty.new("<integration-key>")         # <Pagerduty::EventsApiV1:…>
incident = pagerduty.trigger("<incident-description>") # <Pagerduty::EventsApiV1::Incident:…>
incident.acknowledge
incident.resolve
```

Although the methods remain, they return instances of different types. These new instances implement the same interface, however. So people should have little trouble upgrading.

However, as `Pagerduty` is no longer a class (it's a Ruby module now), if it has been used in any custom inheritance trees, this new version will be incompatible.

For this reason, we should bump a major version number when releasing.

### CODE REVIEW

File: spec/pagerduty/legacy_spec.rb [hunk 1]
```diff
@@ -2,21 +2,27 @@
 
 require "spec_helper"
 
-describe Pagerduty do
+RSpec.describe Pagerduty do
```
Comment (orien) [inline spec/pagerduty/legacy_spec.rb]: Here we prove we still support the old gem 2.0 interface by running the original tests. They all still pass bar one.

File: spec/pagerduty/legacy_spec.rb [hunk 2]
```diff
@@ -325,22 +311,24 @@
...
+        #   }
+        #   Then {
+        #     expect(transport).to have_received(:send_payload).with(
+        #       incident_key: "method param incident_key",
+        #       service_key:  "a-test-service-key",
+        #       event_type:   "trigger",
+        #       description:  "description",
+        #     )
+        #   }
+        # end
```
Comment (orien) [inline spec/pagerduty/legacy_spec.rb]: This is the only test that doesn't pass. In this scenario an `ArgumentError` is now raised.

  ```ruby
  incident1 = pagerduty.trigger('first incident', incident_key: 'one') # this'll work fine
  incident2 = incident1.trigger('second incident', incident_key: 'two') # this no-longer works.
  ```

  The difference is in the object we're calling the method on. ie. It's ok to pass an `incident_key` in the `pagerduty.trigger` method. But not ok to pass it into the `incident.trigger` method.

  Instead always use the `pagerduty` object when triggering new incidents (with new incident keys).

  This works:

  ```ruby
  incident1 = pagerduty.trigger('first incident', incident_key: 'one')
  incident2 = pagerduty.trigger('second incident', incident_key: 'two')
  ```

  And this is even better:

  ```ruby
  incident1 = pagerduty.get_incident('one').trigger('first incident')
  incident2 = pagerduty.get_incident('two').trigger('second incident')
  ```

Comment (johnsyweb) [reply spec/pagerduty/legacy_spec.rb:331]: Is it worth adjusting the specs to demonstrate the new behaviour?

File: CHANGELOG.md [hunk 3]
```diff
@@ -10,7 +10,71 @@ The format is based on [Keep a Changelog], and this project adheres to
...
+
+  And this is even better:
+
+  ```ruby
+  incident1 = pagerduty.get_incident('one').trigger('first incident')
+  incident2 = pagerduty.get_incident('two').trigger('second incident')
+  ```
+
 [Unreleased]: https://github.com/envato/pagerduty/compare/v2.1.3...HEAD
+[#64]: https://github.com/envato/pagerduty/pull/64
```
Comment (johnsyweb) [inline CHANGELOG.md:77]: Super clear. Thank you.

File: CHANGELOG.md [hunk 4]
```diff
@@ -10,7 +10,71 @@ The format is based on [Keep a Changelog], and this project adheres to
 
 ## [Unreleased]
 
+### Added
+
+- A new mechanisim for instantiating a pagerduty instance ([#64]).
+
+  ```ruby
+  pagerduty = Pagerduty.build(integration_key: "<my-integration-key>",
+                              api_version:     "1")
```
Comment (johnsyweb) [inline CHANGELOG.md:19]: API version as a string looks a little strange. I would expect to see this:

```suggestion
                              api_version:     2)
```

File: spec/pagerduty/legacy_spec.rb (hunk 2 above)
Comment (orien) [reply spec/pagerduty/legacy_spec.rb:331]: This spec file (`legacy_spec`), is here to demonstrate compatibility with the current version of this gem. I don't think we want to include anything that wasn't a part of the original behaviour.

FYI: this behaviour is covered in the `events_api_v1_spec` ([see here](https://github.com/envato/pagerduty/pull/64/files#diff-9077cb9fdc5ce6c5d468d7b643d1bf1cR318-R321)).

File: CHANGELOG.md (hunk 4 above)
Comment (orien) [reply CHANGELOG.md]: 👍 Both numeric and string values will work. I've added a test for both in 98765c7, and updated the documentation to use numeric values.

**Comment by orien** [comment]
@johnsyweb, @stergiom Thanks for the review. I appreciate it. 🙇

File: spec/pagerduty/legacy_spec.rb (hunk 2 above)
Comment (johnsyweb) [reply spec/pagerduty/legacy_spec.rb:331]: 💯

**Comment by gaffneyc** [comment]
I realize the question is late but just noticed it now that 3.0 is out. Any reason to introduce `build` rather than reimplement `new` to return a client for the requested version?

```ruby
module Pagerduty
  def self.new(key, options = {})
    version = options.fetch(:api_version, 1)
    events_api_class(version).new(key, options)
  end
end

**Comment by orien** [comment]
Hi @gaffneyc, thanks for the question.

I hope this change hasn't caused problems for you.

I did consider the implementation you've suggested and made following observation.

```ruby
pagerduty = Pagerduty.new(integration_key, api_version: 1)
```

When developers look at this code they'll usually make the following assumptions:

- `Pagerduty` is a class
- `Pagerduty.new` returns an instance of `Pagerduty`

Both of which do not hold.

As a result, I felt uncomfortable baking into the interface such a potential for surprises and confusion.

**Pull request merged**